src/LocalMesh.C

```diff
@@ -334,32 +326,28 @@ void LocalMesh::UpdateFringes(double *qh, double *qd)
 // doesn't work CUDA-Aware -- debug this
 void LocalMesh::UpdateFringes(double *qd)
 {
-    nthreads=device2host.size();
-    if(nthreads == 0) return;
-    n_blocks=nthreads/block_size + (nthreads%block_size==0 ? 0:1);
-    FVSAND_GPU_LAUNCH_FUNC(updateHost,n_blocks,block_size,0,0,
-			   qbuf_d,qd,device2host_d,device2host.size());
-
+    // sanity checks
+    assert( device2host.size() == host2device.size() );
     
+    const int N = static_cast< int >( device2host.size() );
+    if( N  == 0 ) return;
+
+    // pack MPI buffers
+    FVSAND_GPU_KERNEL_LAUNCH( updateHost, N, qbuf_d, qd, device2host_d, N );
 
     // separate sends and receives so that we can overlap comm and calculation
     // in the residual and iteration loops.
-    // TODO (george) use qbuf2_d and qbuf_d instead of qbuf2 and qbuf for cuda-aware
     int reqcount=0;
-    pc.postRecvs_direct(qbuf2,nfields_d,rcvmap,ireq,mycomm,&reqcount);
-    // TODO (george) with cuda-aware this pull is not required
-    // but it doesn't work now
-    gpu::pull_from_device<double>(qbuf,qbuf_d,sizeof(double)*device2host.size());
-    pc.postSends_direct(qbuf,nfields_d,sndmap,ireq,mycomm,&reqcount);
+    pc.postRecvs_direct(qbuf_d2,nfields_d,rcvmap,ireq,mycomm,&reqcount);
+
+    // wait for pack kernel to finish packing buffers?
+    FVSAND::gpu::synchronize();
```

Comment:
Note: I think we need to synchronize here since the pack kernel, `updateHost` in line 336 may have not finished packing the send buffers in `qbuf_d`